src/gadgets/ecc/gate.rs

```diff
@@ -0,0 +1,184 @@
+        condition: &AssignedValue<F>,
+    ) -> Result<AssignedValue<F>, Halo2PlonkError>;
+
+    fn is_infinity_point(
+        &self,
+        ctx: &mut RegionCtx<'_, F>,
+        x: &AssignedValue<F>,
+        y: &AssignedValue<F>,
+    ) -> Result<AssignedValue<F>, Halo2PlonkError>;
+
+    fn is_equal_term(
+        &self,
+        ctx: &mut RegionCtx<'_, F>,
+        a: &AssignedValue<F>,
+        b: &AssignedValue<F>,
+    ) -> Result<AssignedValue<F>, Halo2PlonkError>;
+
+    /// # Safety
+    ///
+    /// TODO
+    unsafe fn unchecked_add<C: CurveAffine<Base = F>>(
+        &self,
+        ctx: &mut RegionCtx<'_, C::Base>,
+        p: &AssignedPoint<C>,
+        q: &AssignedPoint<C>,
+    ) -> Result<AssignedPoint<C>, Halo2PlonkError>;
+
+    /// # Safety
+    ///
+    /// TODO
+    unsafe fn unchecked_double<C: CurveAffine<Base = F>>(
+        &self,
+        ctx: &mut RegionCtx<'_, C::Base>,
+        p: &AssignedPoint<C>,
+    ) -> Result<AssignedPoint<C>, Halo2PlonkError>;
+
+    fn negate<C: CurveAffine<Base = F>>(
+        &self,
+        ctx: &mut RegionCtx<'_, F>,
+        p: &AssignedPoint<C>,
+    ) -> Result<AssignedPoint<C>, Halo2PlonkError>;
+}
+
+impl<const T: usize, F: PrimeField> EccGate<F> for MainGate<F, T> {
+    fn new(config: Self::Config) -> Self {
+        MainGate::new(config)
+    }
+
+    fn assign_point<C: CurveAffine<Base = F>, AN: Into<String>>(
+        &self,
+        ctx: &mut RegionCtx<'_, C::Base>,
+        annotation: impl Fn() -> AN,
+        coords: Option<(C::Base, C::Base)>,
+    ) -> Result<AssignedPoint<C>, Halo2PlonkError> {
+        let x = ctx.assign_advice(
+            || format!("{}.x", annotation().into()),
+            self.config().state[0],
+            Value::known(coords.map_or(C::Base::ZERO, |c| c.0)),
+        )?;
+        let y = ctx.assign_advice(
+            || format!("{}.y", annotation().into()),
+            self.config().state[1],
+            Value::known(coords.map_or(C::Base::ZERO, |c| c.1)),
+        )?;
+        ctx.next();
+
+        Ok(AssignedPoint { x, y })
+    }
+
+    fn conditional_select(
+        &self,
+        ctx: &mut RegionCtx<'_, F>,
+        a: &AssignedValue<F>,
+        b: &AssignedValue<F>,
+        cond: &AssignedValue<F>,
+    ) -> Result<AssignedValue<F>, Halo2PlonkError> {
+        MainGate::conditional_select(self, ctx, a, b, cond)
+    }
+
+    fn is_infinity_point(
+        &self,
+        ctx: &mut RegionCtx<'_, F>,
+        x: &AssignedValue<F>,
+        y: &AssignedValue<F>,
+    ) -> Result<AssignedValue<F>, Halo2PlonkError> {
+        MainGate::is_infinity_point(self, ctx, x, y)
+    }
+
+    fn is_equal_term(
+        &self,
+        ctx: &mut RegionCtx<'_, F>,
+        a: &AssignedValue<F>,
+        b: &AssignedValue<F>,
+    ) -> Result<AssignedValue<F>, Halo2PlonkError> {
+        MainGate::is_equal_term(self, ctx, a, b)
+    }
+
+    /// # Safety
+    ///
+    /// TODO
+    unsafe fn unchecked_add<C: CurveAffine<Base = F>>(
```

Comment:
From https://doc.rust-lang.org/book/ch19-01-unsafe-rust.html

> - Dereference a raw pointer
> - Call an unsafe function or method
> - Access or modify a mutable static variable
> - Implement an unsafe trait
> - Access fields of a union

In general, novice Rust developers are afraid to use `unsafe`, because they think it disables any checks and even borrow-checker by default. However, `unsafe` is actually a simple annotation to the developer "pay attention to the documentation, it will tell you what things you need to check by hand". This could be about memory, any direct access guarantees, etc.

In this case, using unsafe is similar to using `unsafe trait`, where certain invariants we have to check by hand

Reply:
> where certain invariants we have to check by hand

What kind of invariants have to check by hand?